examples/movement/physics_in_fixed_timestep.rs

```diff
@@ -142,55 +184,187 @@ fn spawn_player(mut commands: Commands, asset_server: Res<AssetServer>) {
-            input.y -= 1.0;
-        }
-        if keyboard_input.pressed(KeyCode::KeyA) {
-            input.x -= 1.0;
-        }
-        if keyboard_input.pressed(KeyCode::KeyD) {
-            input.x += 1.0;
-        }
-
-        // Need to normalize and scale because otherwise
-        // diagonal movement would be faster than horizontal or vertical movement.
-        // This effectively averages the accumulated input.
-        velocity.0 = input.extend(0.0).normalize_or_zero() * SPEED;
+    /// Since Bevy's 3D renderer assumes SI units, this has the unit of meters per second.
+    /// Note that about 1.5 is the average walking speed of a human.
+    const SPEED: f32 = 4.0;
+    let (mut input, mut velocity) = player.into_inner();
+    // Reset the input to zero before reading the new input. As mentioned above, we can only do this
+    // because this is continuously pressed by the user. Do not reset e.g. whether the user wants to boost.
+    input.movement = Vec2::ZERO;
+    if keyboard_input.pressed(KeyCode::KeyW) {
+        input.movement.y += 1.0;
+    }
+    if keyboard_input.pressed(KeyCode::KeyS) {
+        input.movement.y -= 1.0;
+    }
+    if keyboard_input.pressed(KeyCode::KeyA) {
+        input.movement.x -= 1.0;
     }
+    if keyboard_input.pressed(KeyCode::KeyD) {
+        input.movement.x += 1.0;
+    }
+
+    // Remap the 2D input to Bevy's 3D coordinate system.
+    // Pressing W makes `input.y` go up. Since Bevy assumes that -Z is forward, we make our new Z equal to -input.y
+    let input_3d = Vec3 {
+        x: input.movement.x,
+        y: 0.0,
+        z: -input.movement.y,
+    };
+
+    // Rotate the input so that forward is aligned with the camera's forward direction.
+    let rotated_input = camera.rotation * input_3d;
+
+    // We need to normalize and scale because otherwise
+    // diagonal movement would be faster than horizontal or vertical movement.
+    // We use `clamp_length_max` instead of `.normalize_or_zero()` because gamepad input
+    // may be smaller than 1.0 when the player is pushing the stick just a little bit.
+    velocity.0 = rotated_input.clamp_length_max(1.0) * SPEED;
+}
+
+/// A simple resource that tells us whether the fixed timestep ran this frame.
+#[derive(Resource, Debug, Deref, DerefMut, Default)]
+pub struct DidFixedTimestepRunThisFrame(bool);
+
+/// Reset the flag at the start of every frame.
+fn clear_fixed_timestep_flag(
+    mut did_fixed_timestep_run_this_frame: ResMut<DidFixedTimestepRunThisFrame>,
+) {
+    did_fixed_timestep_run_this_frame.0 = false;
+}
+
+/// Set the flag during each fixed timestep.
+fn set_fixed_time_step_flag(
+    mut did_fixed_timestep_run_this_frame: ResMut<DidFixedTimestepRunThisFrame>,
+) {
+    did_fixed_timestep_run_this_frame.0 = true;
+}
+
+fn did_fixed_timestep_run_this_frame(
+    did_fixed_timestep_run_this_frame: Res<DidFixedTimestepRunThisFrame>,
+) -> bool {
+    did_fixed_timestep_run_this_frame.0
+}
```

Comment:
This feels like something that should be built in to bevy instead? Seems pretty widely useful.

Reply:
Agreed lol.
Can PR afterwards